csi/src/mount.rs

```diff
@@ -111,58 +111,97 @@ pub fn mount_opts_compare(m1: &[String], m2: &[String], ro: bool) -> bool {
+                name: (*fsname).to_string(),
+                defaults: opts,
+            }),
+            Err(err) => {
+                warn!("Filesystem {} will not be available: {}", fsname, err)
+            }
+        }
+    }
+
+    filesystems
+}
+
+fn probe_filesystem(fsname: &str) -> Result<Vec<String>, String> {
+    let mi = MountInfo {
+        source: format!("/tmp/fs-{}.img", fsname),
+        dest: format!("/tmp/fs_default_{}", fsname),
+        opts: Vec::new(),
+    };
+    prepare_mount(&mi)?;
+
+    match probe_defaults(fsname, &mi) {
+        Ok(opts) => {
+            cleanup_mount(&mi).unwrap_or_else(|e| {
+                warn!("{}", e);
+            });
+            Ok(opts)
+        }
+        Err(err) => Err(format!(
+            "Filesystem {} will not be available: {}",
+            fsname, err
+        )),
+    }
+}
+
+fn prepare_mount(mi: &MountInfo) -> Result<(), String> {
     // im using run_script as its a very nice macro to not "exec" something
     // rather it actually builds a proper script.
     let mut options = ScriptOptions::new();
     options.exit_on_error = true;
 
-    // truncate in busybox does not support units, so be express the size in
+    // truncate in busybox does not support units, so we express the size in
     // bytes
     let (code, _stdout, stderr) = run_script!(
         r#"
-        truncate -s 67108864 /tmp/fs.img
-        mkdir -p /tmp/fs_default
+        truncate -s 67108864 $1
+        mkdir -p $2
         "#,
+        vec!(mi.source.clone(), mi.dest.clone()),
         &options
     )
     .unwrap();
-    if code != 0 {
-        return Err(format!(
-            "Failed to set up default mount options files: {}",
-            stderr
-        ));
-    }
 
-    for fsname in supported_fs.iter() {
-        match probe_defaults(fsname) {
-            Ok(opts) => filesystems.push(Fs {
-                name: (*fsname).to_string(),
-                defaults: opts,
-            }),
-            Err(err) => {
-                warn!("Filesystem {} will not be available: {}", fsname, err)
-            }
-        }
+    if code == 0 {
+        return Ok(());
     }
+    Err(format!(
+        "Failed to set up default mount options files: {}",
+        stderr
+    ))
+}
+
+fn cleanup_mount(mi: &MountInfo) -> Result<(), String> {
+    let mut options = ScriptOptions::new();
+    options.exit_on_error = true;
 
     // stdout of clean up script is always empty
     let (code, _stdout, stderr) = run_script!(
         r#"
-        rmdir /tmp/fs_default
-        rm -f /tmp/fs.img
+        rmdir $2
+        rm -f $1
         "#,
+        &vec!(mi.source.clone(), mi.dest.clone()),
         &options
     )
     .unwrap();
-    if code != 0 {
-        error!("Failed to cleanup default mount options files: {}", stderr);
-    }
 
-    Ok(filesystems)
+    if code == 0 {
+        return Ok(());
+    }
+    Err(format!(
+        "Failed to cleanup default mount options files: {}",
+        stderr
+    ))
```

Comment:
thanks for doing this! It will make the patch for lxd cluster much easier than before.